### ISSUE TITLE
Field composition vs inheritance 

### ISSUE THREAD

**Issue by kettanaito** [kettanaito/react-advanced-form#132]
## What
Need to try to rewrite fields instances to composites based on Field class, rather than its extensions.

## Why
There is a strong argument to prefer composition over inheritance in React. Need to consider that.

## How
1. Render Field within render
1. Separate render logic of Field into a pure function and reuse it in field instances
1. Think of field lifecycle methods

**Comment by kettanaito** [comment]
1. Field HoC using the render of wrapped component as renderElement result.
1. Field providing callback props to the custom field instance.

**Comment by kettanaito** [comment]
Basics.

Need to test custom event handler methods override.

**Comment by kettanaito** [comment]
> **Note to myself:** Don't pass already known props to `WrappedComponent`, just render it and assign new props from Field and context.

**Comment by kettanaito** [comment]
I've opened an intermediate pull request for the review.

Please, @Vidlec, may I ask you to look into it? I am mainly concerned about the methods names and overall components' architecture. 

It's been a breaking change, and I've settled with the `createField` HOC instead of direct composition like so:
```jsx
// This is how it's NOT going to work
export default class Input extends React.Component {
  render() {
    return (<Field component={ input } onRegister={ ... } />);
  }
} 
```

Instead, see the new code for the examples of field composition using `createField` HOC. I will update the docs respectively in the nearest future, before merging the pull request. Thanks.

**Comment by kettanaito** [comment]
Overall, changes are approved, yet it is still uncertain how to pass custom event handlers of `onChange` and other field's events.

1. Provide event handlers on `createField()` level. **Pros:** Should work. **Cons:** Feels and is turned out of the field's scope, very unnatural.
2. Override event handler prop on WrappedComponent.render method:

```jsx
class MyInput extends React.Component {
  handleChange = () => {
    // reads nice, how to pass needed props/methods? Smth like remapProps on createField?
  }

  render() {
    return (<input { ...this.props } onChange={ this.handleChange } />);
  }
}

export default createField()(MyInput);
```

**Comment by kettanaito** [comment]
The release of #126 is planned for this weekend.

This change, although major, should not affect any other behavior of RAF. Therefore the remaining bugs will be covered upon the new field architecture.

**Comment by kettanaito** [comment]
The field composition changes have been merged and are available from [`0.16.0`](https://github.com/kettanaito/react-advanced-form/commit/11deeea249389ad513daac2f1dd6963dffb3545c) release.

I will ensure documentation reflects the new changes properly and close this issue. Hurray.

**Comment by kettanaito** [comment]
## Update
The `0.16.0` release has been published containing this breaking architecture change. Native fields have been re-written.

@jonasdumas please refer to the [`createField()` documentation](https://kettanaito.gitbooks.io/react-advanced-form/hoc/createField.html) on how to integrate a third-party form elements to work with RAF. I have tried to create the interface which would be dynamic and flexible, so all the custom needs may be covered.

Let me know if anything is unclear. Closing.
